### ISSUE TITLE
[Bug] ShadowPEFT LM-head trainability freeze after set_adapter — modules_to_save=["lm_head"] silently stops receiving grads

### ISSUE THREAD

**Issue by Sravanjangam** [huggingface/peft#3626]
### Summary

ShadowPEFT (`#3354`, `peft@9c16ee66`) silently stops training `lm_head` after the first `set_adapter` when using the documented `ShadowConfig(task_type="CAUSAL_LM", modules_to_save=["lm_head"])` workflow. Loss still decreases (adapter trains), so the bug is invisible until eval — `lm_head` was frozen the whole time.

Thanks for ShadowPEFT — the design is strong, this is a small trainability-sync gap. Filing for tracking; happy to PR the 5-line fix after a nod.

---

### Affected code

* `src/peft/tuners/shadow/model.py:731-749` — `_sync_shadow_module_trainability` walks only `self.shadow_backbone / shadow_projection / shadow_head` to flip `requires_grad`.
* `src/peft/tuners/shadow/model.py:787-793` — `_resolve_shadow_head` returns `self.model.get_output_embeddings()` (the base `lm_head`) for `task_type=="CAUSAL_LM"` — by design `shadow_head` is `None` and the base head is reused.
* `src/peft/tuners/shadow/model.py:751-753` — `_mark_only_adapters_as_trainable` delegates to the above sync; `set_adapter:819-825` also calls it.

`lm_head` lives on `self.model` (the base model), not in any `shadow_*` container, so it is never visited by the sync.

---

### Root cause

For `CAUSAL_LM`, ShadowPEFT intentionally does not create a `shadow_head` — it reuses the frozen base `lm_head` and lets users train it via PEFT’s normal `modules_to_save=["lm_head"]` mechanism (docstring `model.py:491-497`, test `tests/test_shadow.py:295-308`). 

`_sync_shadow_module_trainability` was added to fix the fact that `shadow_backbone/projection/head` live on the *tuner* (not on `self.model`), so `BaseTuner.set_adapter`’s walk never reaches them (`model.py:731-738` comment). It correctly syncs those three containers, but it does not sync the base `lm_head` that `_resolve_shadow_head` returns. After the first `set_adapter("other")` or `set_adapter("default", inference_mode=True)`, the tuner’s shadow modules are flipped but `lm_head` stays in its previous `requires_grad` state — frozen when it should be trainable, or trainable when it should be frozen.

---

### Why this is dangerous — silent correctness bug

* **Appears to train:** `loss` decreases normally (adapter gradients flow), so no error or warning.
* **But `lm_head` is frozen:**

| Parameter | Before `set_adapter` | After `set_adapter("other")` (bug present) | After fix |
|-----------|----------------------|---------------------------------------------|-----------|
| `lm_head.weight.grad` | `Tensor` (norm ≈2.43) | `None` | `Tensor` |
| `lm_head.weight.requires_grad` | `True` | `False` (wrong) | `True` |
| Gradient norm | 2.43 | 0 | 2.43 |

* **Eval is wrong:** Checkpoint saves `lm_head` via `modules_to_save` path, but the saved head was never updated — fine-tuning does nothing to the head, yet the user believes it did. This is exactly the kind of bug that survives for months because training “succeeds.”

* **Who can trigger:** Every `ShadowConfig(CAUSAL_LM, modules_to_save=["lm_head"])` workflow that calls `set_adapter` (multi-adapter training, or `inference_mode` toggling). Single-adapter, no `set_adapter` is fine — which is why `test_save_includes_trainable_lm_head` (`test_shadow.py:295`) passes (it never calls `set_adapter`).

---

### Minimal reproduction (CPU)

```python
from transformers import LlamaConfig, LlamaForCausalLM
from peft import get_peft_model, ShadowConfig
import torch

cfg = LlamaConfig(vocab_size=1000, hidden_size=256, intermediate_size=512,
                  num_hidden_layers=4, num_attention_heads=8, num_key_value_heads=8)
base = LlamaForCausalLM(cfg)
model = get_peft_model(base, ShadowConfig(task_type="CAUSAL_LM",
                                          modules_to_save=["lm_head"],
                                          shadow_num_hidden_layers=2))
ids = torch.randint(0, 1000, (2, 8))

# lm_head is trainable initially
print("initial:", any(p.requires_grad for p in model.get_output_embeddings().parameters()))  # True
model(ids, labels=ids).loss.backward()
print("grad before:", model.get_output_embeddings().weight.grad is not None)  # True
model.zero_grad()

# Add second adapter and switch — this is the documented multi-adapter path
model.add_adapter("other", ShadowConfig(task_type="CAUSAL_LM", shadow_num_hidden_layers=2))
model.set_adapter("other")
print("after set_adapter other, requires_grad:", model.get_output_embeddings().weight.requires_grad)  # BUG: False (should be True)

# Train on other
model(ids, labels=ids).loss.backward()
print("grad after switch:", model.get_output_embeddings().weight.grad)  # BUG: None (should be Tensor)
print("grad norm:", 0 if model.get_output_embeddings().weight.grad is None else float(model.get_output_embeddings().weight.grad.norm()))
```

---

### Proposed fix (minimal, very low risk)

In `_sync_shadow_module_trainability` (`model.py:731-749`), also sync the base `lm_head` when it is the resolved shadow head:

```python
def _sync_shadow_module_trainability(self, inference_mode=False):
    self._ensure_shadow_containers()
    for container in (self.shadow_backbone, self.shadow_projection, self.shadow_head):
        for adapter_name, module in container.items():
            module.requires_grad_(adapter_name in self.active_adapters and not inference_mode)
    # --- add ---
    if any(c.task_type == TaskType.CAUSAL_LM for c in self.peft_config.values()):
        # CAUSAL_LM reuses base lm_head via _resolve_shadow_head; it lives outside shadow_* containers
        head = self.model.get_output_embeddings()
        if isinstance(head, torch.nn.Module):
            # Only flip if lm_head was requested via modules_to_save (otherwise leave frozen)
            if any("lm_head" in (c.modules_to_save or []) for c in self.peft_config.values()):
                for p in head.parameters():
                    # trainable when any CAUSAL_LM adapter is active and not inference_mode
                    p.requires_grad = bool(self.active_adapters) and not inference_mode
    # --- end add ---
    for adapter_name, backbone in self.shadow_backbone.items():
        if self.peft_config[adapter_name].shadow_model != "mirror":
            self._freeze_backbone_embeddings(backbone)
```

*No API change, no new config, only fixes the missing walk. Single-adapter, no-`lm_head` runs unchanged.*

Est. diff: 5–8 lines in `model.py:731-749`.

---

### Expected impact

* Correctness for every `CAUSAL_LM` ShadowPEFT run with `modules_to_save=["lm_head"]` that uses `set_adapter` or `inference_mode`.
* Very low migration risk — currently `lm_head` is *always* frozen after `set_adapter`, so the fix only makes it trainable when it was documented to be trainable.

---

### Tests to add

* `test_shadow_lm_head_trainable_after_set_adapter` — the repro above as a `pytest` case: assert `lm_head.weight.grad is not None` and `grad.norm() > 0` after `set_adapter("other")`.
* `test_shadow_lm_head_frozen_in_inference_mode` — `set_adapter("default", inference_mode=True)` → `assert not lm_head.weight.requires_grad`.

---

### Environment

* `peft@9c16ee66` (`main` HEAD 2026-08-31, 0 behind `upstream/main`), Python 3.12, torch 2.13 CPU, transformers 5.15, macOS arm64. Read-only audit at `peft@9c16ee66`.
* Found via `github-audit-harness` Wave-2 (ShadowPEFT surface `tuners/shadow/*` + `tests/test_shadow.py:295-308`), principal direct read + `16_shadow.md` (20 findings, P0-3).

Happy to open a PR with the 5-line fix + two tests after a nod — will include `make style` (ruff) and `pytest tests/test_shadow.py -k "lm_head or trainable"` pass counts.

---

cc @BenjaminBossan — please review when you have a moment, happy to hear your suggestions and ready for a PR with tests + `make style` as soon as you give a nod. Thanks again for the great work on PEFT!

**Comment by cananoo** [comment]
I verified this remains open with no linked open PR. I’m preparing a focused fix in _sync_shadow_module_trainability so a CAUSAL_LM output head requested through modules_to_save follows adapter switching and inference_mode, with regression coverage for both trainable and frozen states. I’ll open a PR after running the targeted ShadowPEFT tests and lint.

**Comment by BenjaminBossan** [comment]
Thanks @Sravanjangam for reporting, feel free to create a PR with the fix.

@cananoo Don't open a PR, the issue author already proposed to provide the fix.

**Comment by BenjaminBossan** [comment]
Closing as discussed on the linked PR.
